### PR TITLE
Add `TensorflowCheckpoint`

### PR DESCRIPTION
**Pull request by bveeramani**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/checkpoint.py [hunk 1]
```diff
@@ -35,108 +35,108 @@
 class Checkpoint:
     """Ray AIR Checkpoint.
 
-    This implementation provides methods to translate between
-    different checkpoint storage locations: Local storage, external storage
-    (e.g. cloud storage), and data dict representations.
+        This implementation provides methods to translate between
```
Comment (ericl) [inline python/ray/air/checkpoint.py:38]: Previous indent looks more correct.

File: python/ray/air/checkpoint.py [hunk 2]
```diff
@@ -202,6 +202,12 @@ def __init__(
         self._uri: Optional[str] = uri
         self._obj_ref: Optional[ray.ObjectRef] = obj_ref
 
+    def _clone_storage_from(self, other: "Checkpoint") -> None:
```
Comment (ericl) [inline python/ray/air/checkpoint.py:205]: Add a doc comment?

**Review by ericl — APPROVED** [review]

File: python/ray/train/tests/test_tensorflow_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    actual_model = checkpoint.get_model(build_linear_model)
+    assert actual_model.get_weights() == expected_model.get_weights()
+
+
+def test_from_model_kwargs():
+    model = build_linear_model()
+    checkpoint = TensorflowCheckpoint.from_model(model, epoch=0)
+    assert checkpoint.to_dict()["epoch"] == 0
+
+
```
Comment (amogkam) [inline python/ray/train/tests/test_tensorflow_checkpoint.py:28]: add test for preprocessor as well?

**Review by amogkam — COMMENTED** [review]
Thanks @bveeramani!

In addition to the comments I left, can we also change the type hints and docstring of `TensorflowPredictor`  to be `TensorflowCheckpoint`?

Also can we add this to the API reference on the docs?

**Comment by amogkam** [comment]
Can you also add a PR description please?

File: python/ray/train/tensorflow/tensorflow_checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+        Returns:
+            A checkpoint that can be loaded by TensorflowPredictor.
+        """
+        checkpoint = Checkpoint.from_dict(
+            {PREPROCESSOR_KEY: preprocessor, MODEL_KEY: model.get_weights(), **kwargs}
+        )
+        return TensorflowCheckpoint(checkpoint)
+
+    def get_model(
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/train/tensorflow/tensorflow_checkpoint.py:43]: get_preprocessor() is defined on the base class (or at least it will in another PR), so it shouldn't be necessary on framework-specific instances

[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/train/tensorflow/tensorflow_checkpoint.py:43]: Is the plan to also fix the dict<>dir compatibility as part of these sequence of PRs? If not, then we'll still run into the same problem of not having a standard format for how the preprocessor is stored in the Checkpoint, right?. Each Checkpoint will have a different implementation on how it extracts the preprocessor.

cc @matthewdeng

[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/train/tensorflow/tensorflow_checkpoint.py:43]: I think the plan is not to fix it, but just do the minimal to ensure preprocessors can always be retrieved no matter the checkpoint format.

**Comment by ericl** [comment]
> In addition to the comments I left, can we also change the type hints and docstring of TensorflowPredictor to be TensorflowCheckpoint?

We shouldn't do this--- let's make sure Predictors always take Checkpoint and not framework-specific types. It would be an interface violation otherwise. TensorflowPredictor can cast internally to TensorFlowCheckpoint if needed.

**Comment by amogkam** [comment]
@ericl can you elaborate a bit more here? what's the downside of only accepting TensorflowCheckpoint? TensorflowPredictor will only work with a TensorflowCheckpoint anyways, so it's better to be explicit?

**Comment by ericl** [comment]
> @ericl can you elaborate a bit more here? what's the downside of only accepting TensorflowCheckpoint? TensorflowPredictor will only work with a TensorflowCheckpoint anyways, so it's better to be explicit?


There is no *advantage* to accepting only TensorFlowCheckpoint.

The disadvantages are:
1. You no longer are implementing the interface correctly. To do this properly, Predictor would have to be Generic[CheckpointCls], which complicates the matter a lot.
2. It's harder to write generic code involving Predictors, since now you need to know their type. This would be a huge headache for BatchPredictor for example.

**Comment by amogkam** [comment]
The advantage is that implicit behavior that users need to be aware of can now be made explicit. We should make it explicit that TorchPredictor cannot work with TFCheckpoint for example.

1. This shouldn't need to change- these framework specific checkpoints subclass ray.air.Checkpoint.
2. I don't think anything needs to change here either. The checkpoint can still be passed through from BatchPredictor to the underlying predictor_cls.

**Comment by amogkam** [comment]
We have to start with what makes sense from the user perspective, and then change our implementation/abstractions to match that.

**Comment by ericl** [comment]
BatchPredictor cannot just pass a Checkpoint class to TensorFlowPredictor. One or the other would have to call `TensorFlowCheckpoint(checkpoint)`. BatchPredictor can't do it since it doesn't know the type of the framework it's calling. TensorFlowPredictor would have to do it internally. But if it's doing it internally, shouldn't the right signature be `from_checkpoint(Checkpoint)` instead of `from_checkpoint(TensorFlowCheckpoint)`?

You could imagine working around this by exposing a `get_checkpoint_cls` method on Predictor, so that BatchPredictor can call `predictor_cls.from_checkpoint(predictor_cls.get_checkpoint_cls()(checkpoint))`, or we can keep it simple and accept Checkpoint.

**Comment by ericl** [comment]
Btw, another issue is that other classes like Trainer are returning `Checkpoint` directly as part of ResultGrid / Result. That would mean you could no longer pass these directly to a Predictor without casting.

**Comment by amogkam** [comment]
yes the latter option will work. The other option that @bveeramani and I were discussing is the checkpoint_cls is stored in the checkpoint data itself so it can always be recovered.

Basically what I want to avoid is something like this 

```
checkpoint = Checkpoint.from_directory(...)
predictor = TorchPredictor.from_checkpoint(checkpoint) # This will fail!
```

This will spit out some error about no MODEL_KEY in dict, which is now internal behavior that we want to hide, and is confusing since I am using directory checkoints.

Instead, if we have the type hint as `TorchCheckpoint`, and in `TorchPredictor` we can raise an error saying if the checkpoint is not of type `TorchCheckpoint`, then this is more clear to the user that I should be doing `TorchCheckpoint.from_model(...)` rather than `Checkpoint.from_directory()`.

**Comment by ericl** [comment]
Storing checkpoint_cls is not a good idea due to Python version compatibility issues (that was the original reason why Predictors weren't stored). How about let's keep the type to Checkpoint for now and revisit with a prototype later? I don't think this is should block this PR.

**Comment by amogkam** [comment]
Discussed offline- for now we will cast to the framework specific class inside the Predictor. More comprehensive changes of internal Checkpoint piping will be punted to future work.

**Comment by amogkam** [comment]
> Storing checkpoint_cls is not a good idea due to Python version compatibility issues

Ah sorry for this I meant storing just the class name as a string, but yes let's revisit this later

File: python/ray/air/checkpoint.py (hunk 1 above)
Comment (ericl) [inline python/ray/air/checkpoint.py:38]: Revert this indentation change.

**Comment by bveeramani** [comment]
Closing in favor #26775. Goofed Git; decided it was easier to start over in a new PR
